python/prima/__init__.py

```diff
@@ -0,0 +1,150 @@
+        else:
+            raise ValueError("Constraint type not recognized")
+
+    if len(nonlinear_constraints) > 0:
+        nonlinear_constraint_function = process_nl_constraints(nonlinear_constraints)
+    else:
+        nonlinear_constraint_function = None
+
+    # Determine if we have multiple linear constraints, just 1, or none, and process accordingly
+    if len(linear_constraints) > 1:
+        linear_constraint = process_multiple_linear_constraints(linear_constraints)
+    elif len(linear_constraints) == 1:
+        linear_constraint = process_single_linear_constraint(linear_constraints[0])
+    else:
+        linear_constraint = None
+
+    return linear_constraint, nonlinear_constraint_function
+
+
+def minimize(fun, x0, args=(), method=None, bounds=None, constraints=(), callback=None, options=None):
+
+    linear_constraint, nonlinear_constraint_function = process_constraints(constraints)
+        
+    quiet = options.get("quiet", True) if options is not None else True
+
+    if method is None:
+        if nonlinear_constraint_function is not None:
+            if not quiet: print("Nonlinear constraints detected, applying COBYLA")
+            method = "cobyla"
+        elif linear_constraint is not None:
+            if not quiet: print("Linear constraints detected without nonlinear constraints, applying LINCOA")
+            method = "lincoa"
+        elif bounds is not None:
+            if not quiet: print("Bounds without linear or nonlinear constraints detected, applying BOBYQA")
+            method = "bobyqa"
+        else:
+            if not quiet: print("No bounds or constraints detected, applying NEWUOA")
+            method = "newuoa"
+    else:
+        # Raise some errors if methods were called with inappropriate options
+        method = method.lower()
+        if method != "cobyla" and nonlinear_constraint_function is not None:
+            raise ValueError("Nonlinear constraints were provided for an algorithm that cannot handle them")
+        if method not in ("cobyla", "lincoa") and linear_constraint is not None:
+            raise ValueError("Linear constraints were provided for an algorithm that cannot handle them")
+        if method not in ("cobyla", "bobyqa", "lincoa") and bounds is not None:
+            raise ValueError("Bounds were provided for an algorithm that cannot handle them")
+
+    try:
+        lenx0 = len(x0)
+    except TypeError:
+        lenx0 = 1
+
+    lb, ub = process_bounds(bounds, lenx0)
+
+    # Project x0 onto the feasible set
+    if nonlinear_constraint_function is None:
+        result = _project(x0, lb, ub, {"linear": linear_constraint, "nonlinear": None})
+        x0 = result.x
+    
+    if linear_constraint is not None:
+        A_eq, b_eq, A_ineq, b_ineq = separate_LC_into_eq_and_ineq(linear_constraint)
+    else:
+        A_eq, b_eq, A_ineq, b_ineq = None, None, None, None
+        
+    if nonlinear_constraint_function is not None:
```

Comment:
Can we move the extraction of `nlconstr0`, `m_nlcon` and `f0` to a lower level function that calls COBYLA, immediately before COBYLA is called? This is needed only by COBYLA.

This is different from reading `lb`, `ub`, ..., `nonlinear_constraint_function`, which can stay here, because they are the information that defines the problem. 

In contrast, `nlconstr0`, `m_nlcon` and `f0` are not part of the problem. They are just an extra piece of  information needed by the Fortran backend of COBYLA due to **limitations of Fortran**. **If COBYLA was implemented using Python or MATLAB, it will not need such information.** 

Putting  the extraction of `nlconstr0`, `m_nlcon` and `f0`  here creats a confusion: it seems that all solvers need to know `f0`. In fact, none of them needs it. (It is evaluated for COBYLA only because we want to keep the evaluation of the constraint and objective functions synchronized.)

Reply:
I had to make some changes to the code since this was written. I have to evaluate the constraint function at the Python level in order to raise exceptions if the lower/upper bounds have a different length from the constraint function output (we must explicitly check for this sort of thing because in a worst case scenario it can lead to the code appearing to execute successfully while ignoring some constraints).

Since the objective and constraint functions are combined at the C++ level, and since I don't want to call the constraint function twice, it is necessary to evaluate `f0` at some point before we go down to the C level. I suppose this could be done in the C++ level, I will make the change.